File: frontend/src/domain/broadcastMessage.ts
import type { NodeRunResult } from '../api/types';

export type BroadcastSeverity = 'info' | 'notice' | 'success';

export interface BroadcastSegment {
    node_id: string;
    body: string;
    severity?: BroadcastSeverity;
    title?: string;
    render_markdown?: boolean;
    source?: string;
    step_number?: number;
}

const MARKDOWN_PATTERNS = [
    /^#{1,6}\s+\S/m,
    /```/,
    /^\s*[-*+]\s+\S/m,
    /^\s*\d+\.\s+\S/m,
    /\*\*[^*\n]+\*\*/,
    /__[^_\n]+__/,
];

export function normalizeBroadcastSeverity(raw: unknown): BroadcastSeverity {
    const s = String(raw ?? 'info').trim().toLowerCase();
    if (s === 'info' || s === 'notice' || s === 'success') return s;
    return 'info';
}

export function looksLikeMarkdown(text: string): boolean {
    if (!text.trim()) return false;
    return MARKDOWN_PATTERNS.some(p => p.test(text));
}

export function broadcastSegmentFromNodeResult(nr: NodeRunResult): BroadcastSegment | null {
    if (nr.status !== 'ok' || !nr.details) return null;
    const det = nr.details as Record<string, unknown>;
    const raw = det.broadcast_segment;
    if (!raw || typeof raw !== 'object') return null;
    const seg = raw as Record<string, unknown>;
    const body = seg.body;
    if (typeof body !== 'string' || !body.trim()) return null;
    return {
        node_id: String(seg.node_id ?? nr.node_id),
        body,
        severity: normalizeBroadcastSeverity(seg.severity),
        title: typeof seg.title === 'string' && seg.title.trim() ? seg.title.trim() : undefined,
        render_markdown:
            typeof seg.render_markdown === 'boolean' ? seg.render_markdown : looksLikeMarkdown(body),
        source: typeof seg.source === 'string' && seg.source.trim() ? seg.source.trim() : undefined,
        step_number: typeof seg.step_number === 'number' ? seg.step_number : nr.step_number,
    };
}

export function collectBroadcastSegmentsFromNodeResults(
    nodeResults: NodeRunResult[] | undefined,
    source?: string,
): BroadcastSegment[] {
    if (!nodeResults?.length) return [];
    const sorted = [...nodeResults].sort((a, b) => {
        const sa = a.step_number ?? 0;
        const sb = b.step_number ?? 0;
        if (sa !== sb) return sa - sb;
        return a.node_id.localeCompare(b.node_id);
    });
    const out: BroadcastSegment[] = [];
    for (const nr of sorted) {
        const seg = broadcastSegmentFromNodeResult(nr);
        if (!seg) continue;
        out.push(source && !seg.source ? { ...seg, source } : seg);
    }
    return out;
}

export function collectBroadcastSegmentsFromRuns(
    runs: Array<{ node_results?: NodeRunResult[] } | null | undefined>,
): BroadcastSegment[] {
    const merged: BroadcastSegment[] = [];
    for (const run of runs) {
        if (!run?.node_results?.length) continue;
        merged.push(...collectBroadcastSegmentsFromNodeResults(run.node_results));
    }
    merged.sort((a, b) => {
        const sa = a.step_number ?? 0;
        const sb = b.step_number ?? 0;
        if (sa !== sb) return sa - sb;
        return a.node_id.localeCompare(b.node_id);
    });
    return merged;
}

export function parseBroadcastSegmentsFromEffects(raw: unknown): BroadcastSegment[] {
    if (!Array.isArray(raw)) return [];
    const out: BroadcastSegment[] = [];
    for (const item of raw) {
        if (!item || typeof item !== 'object') continue;
        const seg = item as Record<string, unknown>;
        const body = seg.body;
        if (typeof body !== 'string' || !body.trim()) continue;
        out.push({
            node_id: String(seg.node_id ?? ''),
            body,
            severity: normalizeBroadcastSeverity(seg.severity),
            title: typeof seg.title === 'string' && seg.title.trim() ? seg.title.trim() : undefined,
            render_markdown:
                typeof seg.render_markdown === 'boolean'
                    ? seg.render_markdown
                    : looksLikeMarkdown(body),
            source: typeof seg.source === 'string' && seg.source.trim() ? seg.source.trim() : undefined,
            step_number: typeof seg.step_number === 'number' ? seg.step_number : undefined,
        });
    }
    return out;
}

export function severityAccentClass(severity: BroadcastSeverity | undefined): string {
    switch (severity ?? 'info') {
        case 'success':
            return 'border-t-success bg-success-muted/30';
        case 'notice':
            return 'border-t-amber-500 bg-amber-500/5 dark:bg-amber-500/10';
        default:
            return 'border-t-mw-primary bg-mw-primary-muted/40';
    }
}

File: backend/app/domain/workflow_executor/broadcast_message.py
"""Broadcast Message utility helpers (segment shape, markdown heuristic, run aggregation)."""

from __future__ import annotations

import re
from typing import Any, Literal, Optional

from app.domain.schemas import NodeRunResult, WorkflowRunResult

BroadcastSeverity = Literal["info", "notice", "success"]

_VALID_SEVERITIES = frozenset({"info", "notice", "success"})

_MARKDOWN_PATTERNS = (
    re.compile(r"^#{1,6}\s+\S", re.MULTILINE),
    re.compile(r"```"),
    re.compile(r"^\s*[-*+]\s+\S", re.MULTILINE),
    re.compile(r"^\s*\d+\.\s+\S", re.MULTILINE),
    re.compile(r"\*\*[^*\n]+\*\*"),
    re.compile(r"__[^_\n]+__"),
)


def normalize_broadcast_severity(raw: Any) -> BroadcastSeverity:
    s = str(raw or "info").strip().lower()
    if s in _VALID_SEVERITIES:
        return s  # type: ignore[return-value]
    return "info"


def looks_like_markdown(text: str) -> bool:
    if not text or not text.strip():
        return False
    return any(p.search(text) for p in _MARKDOWN_PATTERNS)


def build_broadcast_segment(
    *,
    node_id: str,
    body: str,
    title: Optional[str] = None,
    severity: BroadcastSeverity = "info",
    source: Optional[str] = None,
    step_number: Optional[int] = None,
) -> dict[str, Any]:
    title_clean = (title or "").strip() or None
    seg: dict[str, Any] = {
        "node_id": node_id,
        "body": body,
        "severity": severity,
        "render_markdown": looks_like_markdown(body),
    }
    if title_clean:
        seg["title"] = title_clean
    if source:
        seg["source"] = source
    if step_number is not None:
        seg["step_number"] = step_number
    return seg


def broadcast_segment_from_node_result(nr: NodeRunResult) -> dict[str, Any] | None:
    if nr.status != "ok" or not nr.details:
        return None
    det = nr.details if isinstance(nr.details, dict) else {}
    raw = det.get("broadcast_segment")
    if not isinstance(raw, dict):
        return None
    body = raw.get("body")
    if not isinstance(body, str) or not body.strip():
        return None
    return dict(raw)


def collect_broadcast_segments_from_node_results(
    node_results: list[NodeRunResult] | None,
    *,
    source: Optional[str] = None,
) -> list[dict[str, Any]]:
    if not node_results:
        return []
    sorted_rows = sorted(
        node_results,
        key=lambda nr: (nr.step_number if nr.step_number is not None else 0, nr.node_id),
    )
    out: list[dict[str, Any]] = []
    for nr in sorted_rows:
        seg = broadcast_segment_from_node_result(nr)
        if seg is None:
            continue
        if source and "source" not in seg:
            seg = {**seg, "source": source}
        out.append(seg)
    return out


def append_broadcast_segments_from_run(
    target: list[dict[str, Any]],
    run_result: WorkflowRunResult | None,
    *,
    source: Optional[str] = None,
) -> None:
    if run_result is None:
        return
    for seg in collect_broadcast_segments_from_node_results(run_result.node_results, source=source):
        target.append(seg)


def merge_broadcast_segment_lists(*lists: list[dict[str, Any]]) -> list[dict[str, Any]]:
    merged: list[dict[str, Any]] = []
    for lst in lists:
        merged.extend(lst)
    merged.sort(
        key=lambda s: (
            s.get("step_number") if isinstance(s.get("step_number"), int) else 0,
            str(s.get("node_id") or ""),
        )
    )
    return merged
